foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp

```diff
@@ -0,0 +1,571 @@
+
+static const websocketpp::log::level APP = websocketpp::log::alevel::app;
+static const websocketpp::log::level RECOVERABLE = websocketpp::log::elevel::rerror;
+
+struct ChannelWithoutId {
+  std::string topic;
+  std::string encoding;
+  std::string schemaName;
+  std::string schema;
+
+  bool operator==(const ChannelWithoutId& other) const {
+    return topic == other.topic && encoding == other.encoding && schemaName == other.schemaName &&
+           schema == other.schema;
+  }
+};
+
+struct Channel : ChannelWithoutId {
+  ChannelId id;
+
+  explicit Channel(ChannelId id, ChannelWithoutId ch)
+      : ChannelWithoutId(std::move(ch))
+      , id(id) {}
+
+  friend void to_json(json& j, const Channel& channel) {
+    j = {
+      {"id", channel.id},
+      {"topic", channel.topic},
+      {"encoding", channel.encoding},
+      {"schemaName", channel.schemaName},
+      {"schema", channel.schema},
+    };
+  }
+
+  bool operator==(const Channel& other) const {
+    return id == other.id && ChannelWithoutId::operator==(other);
+  }
+};
+
+enum class BinaryOpcode : uint8_t {
+  MESSAGE_DATA = 1,
+};
+
+enum class StatusLevel : uint8_t {
+  INFO = 0,
+  WARNING = 1,
+  ERROR = 2,
+};
+
+template <typename ServerConfiguration>
+class Server final {
+public:
+  using ServerType = websocketpp::server<ServerConfiguration>;
+  using ConnectionType = websocketpp::connection<ServerConfiguration>;
+  using MessagePtr = typename ServerType::message_ptr;
+
+  static const std::string SUPPORTED_SUBPROTOCOL;
+
+  explicit Server(std::string name, LogCallback logger);
+  ~Server();
+
+  Server(const Server&) = delete;
+  Server(Server&&) = delete;
+  Server& operator=(const Server&) = delete;
+  Server& operator=(Server&&) = delete;
+
+  void start(uint16_t port);
+  void stop();
+
+  ChannelId addChannel(ChannelWithoutId channel);
+  void removeChannel(ChannelId chanId);
+  void broadcastChannels();
+
+  void setSubscribeHandler(std::function<void(ChannelId)> handler);
+  void setUnsubscribeHandler(std::function<void(ChannelId)> handler);
+
+  void sendMessage(ChannelId chanId, uint64_t timestamp, std::string_view data);
+
+  std::optional<asio::ip::tcp::endpoint> localEndpoint();
+
+private:
+  struct ClientInfo {
+    std::string name;
+    ConnHandle handle;
+    std::unordered_map<ChannelId, SubscriptionId> subscriptionsByChannel;
+
+    ClientInfo(const ClientInfo&) = delete;
+    ClientInfo& operator=(const ClientInfo&) = delete;
+
+    ClientInfo(ClientInfo&&) = default;
+    ClientInfo& operator=(ClientInfo&&) = default;
+  };
+
+  std::string _name;
+  LogCallback _logger;
+  ServerType _server;
+  std::unique_ptr<std::thread> _serverThread;
+
+  uint32_t _nextChannelId = 0;
+  std::map<ConnHandle, ClientInfo, std::owner_less<>> _clients;
+  std::unordered_map<ChannelId, Channel> _channels;
+  std::function<void(ChannelId)> _subscribeHandler;
+  std::function<void(ChannelId)> _unsubscribeHandler;
+  std::shared_mutex _clientsChannelMutex;
```

Comment:
For the future it probably makes sense to have two separate mutexes, one for channels and one for clients.